packages/slate/src/utils/text-utils.js

```diff
@@ -81,168 +138,144 @@ function getCharLength(char) {
         }
-      } else {
-        if (prev == 'SURR' || prev == 'BMP') {
-          break
-        }
-      }
 
-      offset += 2
-      prev = modifier ? 'MOD' : 'SURR'
-      charCode = text.charCodeAt(offset)
-      // Absolutely fine to `continue` without any
-      // checks because if `charCode` is NaN (which
-      // is the case when out of `text` range), next
-      // `while` loop won't execute and we're done.
-      continue
-    }
+        offset += 1
+        prev = 'BMP'
+        charCode = text.charCodeAt(offset)
+        continue
+      }
 
-    // If zero width joiner
-    if (charCode == 0x200D) {
-      offset += 1
-      prev = 'ZWJ'
-      charCode = text.charCodeAt(offset)
-      continue
-    }
+      if (isVariationSelector(charCode)) {
+        if (!forward && prev && prev !== 'ZWJ') {
+          break
+        }
 
-    if (isBMPEmoji(charCode)) {
-      if (
-        (forward && prev == 'VAR') ||
-        (prev && prev != 'ZWJ' && prev != 'VAR')
-      ) {
-        break
+        offset += 1
+        prev = 'VAR'
+        charCode = text.charCodeAt(offset)
+        continue
       }
-      offset += 1
-      prev = 'BMP'
-      charCode = text.charCodeAt(offset)
-      continue
-    }
 
-    if (isVarSelector(charCode)) {
-      if (!forward && prev && prev != 'ZWJ') {
+      // Modifier "fuses" with what ever character is before that
+      // (even whitespace), need to look ahead if loop gets here.
+      if (forward) {
+        const nextCharCode = text.charCodeAt(offset + 1)
+
+        if (isModifier(nextCharCode, text, offset + 1)) {
+          offset += 3
+          prev = 'MOD'
+          charCode = text.charCodeAt(offset)
+          continue
+        }
+      } else if (prev === 'MOD') {
+        offset += 1
         break
       }
-      offset += 1
-      prev = 'VAR'
-      charCode = text.charCodeAt(offset)
-      continue
-    }
 
-    // Modifier "groups up" with what ever character is
-    // before that (even whitespace), need to look ahead.
-    if (forward) {
-      const nextCharCode = text.charCodeAt(offset + 1)
-      if (isModifier(nextCharCode, offset + 1, text)) {
-        offset += 3
-        prev = 'MOD'
-        charCode = text.charCodeAt(offset)
-        continue
-      }
-    } else if (prev == 'MOD') {
-      offset += 1
+      // If while loop ever gets here, we're
+      // done (e.g Latin characters, length 1).
+      if (prev === null) offset += 1
       break
     }
-
-    // If while loop ever gets here,
-    // we're done (e.g latin chars).
-    break
   }
 
-  return offset || 1
+  return offset
 }
 
 /**
- * Get the offset to the end of the character before an `offset` in `text`.
+ * Get the offset to the end of character(s) before an `offset` in `text`.
  *
  * @param {String} text
  * @param {Number} offset
+ * @param {Number} chars
  * @return {Number}
  */
 
-function getCharOffsetBackward(text, offset) {
+function getCharsOffsetBackward(text, offset, chars) {
```

Comment:
@ianstormtaylor I think it makes sense to also add default `chars = 1` here (and forward function)? Should I retain original `getCharOffsetBackward` and add `getCharsOffsetBackward` as separate function? Even though this API is not documented, it would be a breaking change.

Reply:
Having that argument makes sense to me, good call. Let's just keep the function name non-plural for now? And we can fix it later with a deprecation when needed.